causalml/inference/meta/tmle.py

```diff
@@ -0,0 +1,175 @@
+            y (np.array): an outcome vector
+            segment (np.array, optional): An optional segment vector of int. If given, the ATE and its CI will be
+                                          estimated for each segment.
+            return_ci (bool, optional): Whether to return confidence intervals
+
+        Returns:
+            (tuple): The ATE and its confidence interval (LB, UB) for each treatment, t and segment, s
+        """
+        check_control_in_treatment(treatment, self.control_name)
+        self.t_groups = np.unique(treatment[treatment != self.control_name])
+        self.t_groups.sort()
+
+        check_p_conditions(p, self.t_groups)
+        if isinstance(p, np.ndarray):
+            treatment_name = self.t_groups[0]
+            p = {treatment_name: p}
+
+        ate = []
+        ate_lb = []
+        ate_ub = []
+
+        for i, group in enumerate(self.t_groups):
+            logger.info('Estimating ATE for group {}.'.format(group))
+            w_group = (treatment == group).astype(int)
+            p_group = p[group]
+
+            if self.calibrate_propensity:
+                logger.info('Calibrating propensity scores.')
+                p_group = calibrate(p_group, w_group)
+
+            yhat_c = np.zeros_like(y, dtype=float)
+            yhat_t = np.zeros_like(y, dtype=float)
+            if self.cv:
+                for i_fold, (i_trn, i_val) in enumerate(self.cv.split(X, y), 1):
+                    logger.info('Training an outcome model for CV #{}'.format(i_fold))
+                    self.model_tau.fit(np.hstack((X[i_trn], w_group[i_trn].reshape(-1, 1))), y[i_trn])
+
+                    yhat_c[i_val] = self.model_tau.predict(np.hstack((X[i_val], np.zeros((len(i_val), 1)))))
+                    yhat_t[i_val] = self.model_tau.predict(np.hstack((X[i_val], np.ones((len(i_val), 1)))))
+
+            else:
+                self.model_tau.fit(np.hstack((X, w_group.reshape(-1, 1))), y)
+
+                yhat_c = self.model_tau.predict(np.hstack((X, np.zeros((len(y), 1)))))
+                yhat_t = self.model_tau.predict(np.hstack((X, np.ones((len(y), 1)))))
+
+            if segment is None:
+                logger.info('Training the TMLE learner.')
+                _ate, se = simple_tmle(y, w_group, yhat_c, yhat_t, p_group)
+                _ate_lb = _ate - se * norm.ppf(1 - self.ate_alpha / 2)
+                _ate_ub = _ate + se * norm.ppf(1 - self.ate_alpha / 2)
+            else:
+                assert segment.shape[0] == X.shape[0] and segment.ndim == 1, 'Segment must be the 1-d np.array of int.'
+                segments = np.unique(segment)
+
+                _ate = []
+                _ate_lb = []
+                _ate_ub = []
+                for s in sorted(segments):
+                    logger.info('Training the TMLE learner for segment {}.'.format(s))
+                    filt = (segment == s) & (yhat_c >= 0) & (yhat_t >= 0) & (yhat_c < np.quantile(yhat_c, q=.99))
```

Comment:
The filter on yhat_c>=0 and yhat_t>=0 may not be necessary for general situations.

Reply:
Done!